modules/040-node-manager/hooks/yc_delete_preemptible_instances.go

```diff
@@ -0,0 +1,365 @@
+	if err != nil {
+		return nil, fmt.Errorf("can't access class name of Machine %q: %s", obj.GetName(), err)
+	}
+	if len(classKind) == 0 {
+		return nil, fmt.Errorf("spec.class.kind is empty in %q", obj.GetName())
+	}
+
+	className, _, err := unstructured.NestedString(obj.UnstructuredContent(), "spec", "class", "name")
+	if err != nil {
+		return nil, fmt.Errorf("can't access class name of Machine %q: %s", obj.GetName(), err)
+	}
+	if len(className) == 0 {
+		return nil, fmt.Errorf("spec.class.name is empty in %q", obj.GetName())
+	}
+
+	return &Machine{
+		Name:             obj.GetName(),
+		Terminating:      terminating,
+		MachineClassKind: classKind,
+		MachineClassName: className,
+	}, nil
+}
+
+func applyNodeGroupFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	icKind, icExists, err := unstructured.NestedString(obj.UnstructuredContent(), "spec", "cloudInstances", "classReference", "kind")
+	if err != nil {
+		return nil, fmt.Errorf("cannot access \"spec.cloudInstances.classReference.kind\" in a NodeGroup %s: %s", obj.GetName(), err)
+	}
+
+	if !icExists || (icKind != "YandexInstanceClass") {
+		return nil, nil
+	}
+
+	nodeCountRaw, nodeCountExists, err := unstructured.NestedFieldNoCopy(obj.UnstructuredContent(), "status", "nodes")
+	if err != nil {
+		return nil, fmt.Errorf("cannot access \"status.nodes\" in a NodeGroup %s: %s", obj.GetName(), err)
+	}
+	readyNodeCountRaw, readyNodeCountExists, err := unstructured.NestedFieldNoCopy(obj.UnstructuredContent(), "status", "ready")
+	if err != nil {
+		return nil, fmt.Errorf("cannot access \"status.ready\" in a NodeGroup %s: %s", obj.GetName(), err)
+	}
+
+	if !nodeCountExists || !readyNodeCountExists {
+		return nil, nil
+	}
+
+	var nodeCount, readyNodeCount int64
+	if os.Getenv("D8_IS_TESTS_ENVIRONMENT") != "" {
+		nodeCount = int64(nodeCountRaw.(float64))
+		readyNodeCount = int64(readyNodeCountRaw.(float64))
+	} else {
+		nodeCount = nodeCountRaw.(int64)
+		readyNodeCount = readyNodeCountRaw.(int64)
+	}
+
+	if (nodeCount < 0) || (readyNodeCount < 0) {
+		return nil, nil
+	}
+
+	return &NodeGroupStatus{
+		Name:  obj.GetName(),
+		Nodes: nodeCount,
+		Ready: readyNodeCount,
+	}, nil
+}
+
+func applyNodeFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	labels := obj.GetLabels()
+
+	ng, ok := labels["node.deckhouse.io/group"]
+	if !ok {
+		return nil, nil
+	}
+
+	return &Node{
+		Name:              obj.GetName(),
+		NodeGroup:         ng,
+		CreationTimestamp: obj.GetCreationTimestamp(),
+	}, nil
+}
+
+func isPreemptibleFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	preemptible, ok, err := unstructured.NestedBool(obj.UnstructuredContent(), "spec", "schedulingPolicy", "preemptible")
+	if err != nil {
+		return nil, fmt.Errorf("can't access field \"spec.schedulingPolicy.preemptible\" of YandexMachineClass %q: %s", obj.GetName(), err)
+	}
+
+	if ok && preemptible {
+		return &YandexMachineClass{
```

Comment:
Why is the response wrapped in a struct? It is not required.

Reply:
Switched to string.